scipy/special/cephes/ndtr.c

```diff
@@ -224,6 +224,84 @@ double ndtr(double a)
 }
 
 
+#define NDTR_DELTA_TAIL_LIMIT  38.5
+
+//
+// Compute ndtr(b) - ndtr(a) robustly.
+//
+double ndtr_delta(double a, double b)
+{
+    int sgn = 1;
+    double delta;
+
+    if (cephes_isnan(a) || cephes_isnan(b)) {
+        sf_error("ndtr_delta", SF_ERROR_DOMAIN, NULL);
+        return NPY_NAN;
+    }
+    if (a == b) {
+        return 0.0;
+    }
+    if (a > b) {
+        double tmp = a;
+        a = b;
+        b = tmp;
+        sgn = -1;
+    }
+    if ((a > NDTR_DELTA_TAIL_LIMIT) || (b < -NDTR_DELTA_TAIL_LIMIT)) {
+        return 0.0;
+    }
+    if (a > 0) {
+        delta = ndtr(-a) - ndtr(-b);
+    }
+    else {
+        delta = ndtr(b) - ndtr(a);
+    }
+    if (delta < 0) {
+        delta = 0.0;
+    }
+    return sgn*delta;
+}
+
+//
+// Compute the CDF for the truncated normal distribution.
+//
+double trunc_ndtr(double a, double b, double x)
+{
+    double delta;
+    double logp;
+
+    if (cephes_isnan(x) || cephes_isnan(a) || cephes_isnan(b) || (a >= b)) {
+        sf_error("trunc_ndtr", SF_ERROR_DOMAIN, NULL);
+        return NPY_NAN;
+    }
+    if (x <= a) {
+        return 0.0;
+    }
+    if (x >= b) {
+        return 1.0;
+    }
+    delta = ndtr_delta(b, a);
+    if (delta > 0) {
+        return ndtr_delta(x, a) / delta;
+    }
+    if (a < 0) {
```

Comment:
Originally I thought this code just ran when there was perfect subtractive cancellation, but it seems to run all the time given the order of arguments passed to `ndtr_delta`.